Refactors radix cache for extra key support

JustinTong0323 opened the pull request:
<!-- Thank you for your contribution! Please follow these guidelines to enhance your pull request. If anything is unclear, submit your PR and reach out to maintainers for assistance. Join our Slack community at https://slack.sglang.ai to discuss further. -->

## Motivation
In specific scenarios, it is necessary to isolate the kvcache for certain requests. For instance, requests utilizing different LoRA adapters, or requests for which we do not desire sharing the kvcache with others(ref https://github.com/sgl-project/sglang/issues/9163).

This PR refactors the Key in the Radix cache by adding an `extra_key` field. Only requests/tokens in the cache that have the same `extra_key` can share the prefix KV cache.

I would implement the logic for cache salting in following PR.

<!-- Describe the purpose and goals of this pull request. -->

## Modifications

This commit refactors the radix cache to support an extra key (e.g., lora_id, cache_salt) for classifying requests.

The changes include:

- Adds `extra_key` to the `Req` class 
- Replace raw token_ids key to `BaseKey` class  to facilitate request classification.
- Modifies the `match_prefix` method in `RadixCache` and `SchedulePolicy` to incorporate the `BaseKey` when matching prefixes.
- **Removes the LoRARadixCache**, as the base RadixCache can now handle LoRA IDs through the `extra_key` mechanism.

## TODO List:
- [x] Add support for HiRadixCache
- [x] Support SWA radix cache
- [x] Compatible for cpp/lmc
- [ ] Support radix cache cpp for extra keys (need modify cpp codes, in following PR
- [ ] others I may not considered

<!-- Detail the changes made in this pull request. -->

## Accuracy Tests

<!-- If this pull request affects model outputs (e.g., changes to the kernel or model forward code), provide accuracy test results. -->

## Benchmarking and Profiling

<!-- If this pull request impacts inference speed, provide benchmarking and profiling results. -->

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

JustinTong0323 committed:
Refactors radix cache for extra key support

This commit refactors the radix cache to support an extra key
(e.g., lora_id, cache_salt) for classifying requests.

The changes include:

- Adds `extra_key` to the `Req` class and `BaseKey` class to
  facilitate request classification.
- Modifies the `match_prefix` method in `RadixCache` and
  `SchedulePolicy` to incorporate the `extra_key` when matching
  prefixes.
- Removes the LoRARadixCache, as the base RadixCache can now
  handle LoRA IDs through the `extra_key` mechanism.

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
support hicache

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
lmcache radix compatible

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
cpp radix cache compatible (not support extra_key kv cache isolation

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
swa radix cache support

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
Merge branch 'main' into feat-refactor-radix-cache-combine-lora-cache

ispobock commented on python/sglang/srt/mem_cache/radix_cache.py:
```diff
@@ -40,14 +40,38 @@
     from sglang.srt.managers.schedule_batch import Req
 
 
+class BaseKey:
```
@xiezhq-hermann what do you think?

JustinTong0323 committed:
Fixes BaseKey indexing for single tokens

Fixes an issue where indexing a BaseKey with a single token was returning an integer instead of a BaseKey object.

This change ensures that even when indexing with a single token, a BaseKey object is returned, maintaining consistency and preventing potential errors.

Adds a unit test for radix cache to improve code coverage.

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
Merge branch 'main' into feat-refactor-radix-cache-combine-lora-cache

JustinTong0323 committed:
Simplifies key handling in radix cache

Updates radix cache implementations to directly use the key object for slicing,
rather than accessing the underlying token_ids and extra_key attributes.
This simplifies the code and improves readability.

Also, initializes the value tensor with the correct data type.
Adds an advanced prefix match test.

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
code clean up

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 commented on test/srt/test_swa_unittest.py:
```diff
@@ -4,7 +4,8 @@
 from sglang.srt.mem_cache.allocator import SWAKVPool, SWATokenToKVPoolAllocator
 from sglang.srt.mem_cache.memory_pool import ReqToTokenPool
-from sglang.srt.mem_cache.radix_cache import SWARadixCache
+from sglang.srt.mem_cache.radix_cache import BaseKey
```
Added.

JustinTong0323 commented on python/sglang/srt/mem_cache/radix_cache.py:
```diff
@@ -192,12 +236,12 @@ def match_prefix(self, key: List[int], **kwargs) -> MatchResult:
 
         if value is None:
-            value = [x for x in key]
+            value = torch.tensor([x for x in key.token_ids], dtype=torch.int64)
```
but it seems not used... not sure if we could directly delete this `if value is None:`

Fridge003 commented:
@JustinTong0323 Please fix the conflict

Fridge003 commented:
I feel the key of this PR is to make sure it doesn't break all kinds of features (hi-cache, swa-radix...) So please make sure this PR isn't merged until all the related tests can pass

JustinTong0323 committed:
Merge branch 'main' into feat-refactor-radix-cache-combine-lora-cache

JustinTong0323 committed:
Enhance match_prefix documentation for clarity and detail

JustinTong0323 committed:
code cleanup

JustinTong0323 committed:
remove duplicate code

JustinTong0323 committed:
upd

Fridge003 reviewed (approved):
LGTM

xiezhq-hermann reviewed (approved):
Looks good to me, would suggest to rename `BaseKey` to `RadixKey` though

xiezhq-hermann committed:
Merge branch 'main' into feat-refactor-radix-cache-combine-lora-cache

JustinTong0323 committed:
Rename "BaseKey" -> "RadixKey"

JustinTong0323 committed:
Merge branch 'main' into feat-refactor-radix-cache-combine-lora-cache
